lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+package drivers
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/canonical/lxd/lxd/db"
+	"github.com/canonical/lxd/lxd/locking"
+	"github.com/canonical/lxd/lxd/revert"
+	"github.com/canonical/lxd/lxd/util"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// powerFlexBlockVolSuffix suffix used for block content type volumes.
+const powerFlexBlockVolSuffix = ".block"
+
+// powerFlexISOVolSuffix suffix used for iso content type volumes.
+const powerFlexISOVolSuffix = ".iso"
+
+// powerFlexVolTypePrefixes maps volume type to storage volume name prefix.
+var powerFlexVolTypePrefixes = map[VolumeType]string{
+	VolumeTypeContainer: db.StoragePoolVolumeTypeNameContainer,
+	VolumeTypeVM:        db.StoragePoolVolumeTypeNameVM,
+	VolumeTypeImage:     db.StoragePoolVolumeTypeNameImage,
+	VolumeTypeCustom:    db.StoragePoolVolumeTypeNameCustom,
+}
+
+// powerFlexError contains arbitrary error responses from PowerFlex.
+type powerFlexError map[string]any
+
+// tokenFactory holds an access token together with a read write mutex.
+type tokenFactory struct {
+	token string
+	mutex *sync.RWMutex
+}
+
+// powerFlexClient holds the PowerFlex HTTP client and an access token factory.
+type powerFlexClient struct {
+	driver *powerflex
+	token  *tokenFactory
+}
+
+// powerFlexStoragePool represents a storage pool in PowerFlex.
+type powerFlexStoragePool struct {
+	ID                 string `json:"id"`
+	Name               string `json:"name"`
+	ProtectionDomainID string `json:"protectionDomainId"`
+}
+
+// powerFlexProtectionDomain represents a protection domain in PowerFlex.
+type powerFlexProtectionDomain struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	SystemID string `json:"systemId"`
+}
+
+// powerFlexProtectionDomainSdsRelation represents an SDS related to a protection domain in PowerFlex.
+type powerFlexProtectionDomainSdsRelation struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	IPList []struct {
+		IP string `json:"ip"`
+	} `json:"ipList"`
+}
+
+// powerFlexSdc represents a SDC in PowerFlex.
+type powerFlexSdc struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	HostType string `json:"hostType"`
+	NQN      string `json:"nqn"`
+}
+
+// powerFlexVolume represents a volume in PowerFlex.
+type powerFlexVolume struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	VolumeType    string `json:"volumeType"`
+	MappedSdcInfo []struct {
+		SdcID    string `json:"sdcId"`
+		SdcName  string `json:"sdcName"`
+		NQN      string `json:"nqn"`
+		HostType string `json:"hostType"`
+	} `json:"mappedSdcInfo"`
+}
+
+// Error tries to return all kinds of errors from the PowerFlex API in a nicely formatted way.
+func (p *powerFlexError) Error() string {
```

Comment:
Can you move this up to where the powerFlexError type is define to keep the types and their functions together?

Reply:
Moved.